BuildScript/BuildTasks/TestConfigurationsFactory.cs

```diff
@@ -0,0 +1,271 @@
+      _supportedBrowsers = supportedBrowsers;
+      _supportedDatabaseSystems = supportedDatabaseSystems;
+      _supportedPlatforms = supportedPlatforms;
+      _supportedConfigurationIDs = supportedConfigurationIDs;
+      _supportedExecutionRuntimes = supportedExecutionRuntimes;
+      _supportedTargetRuntimes = supportedTargetRuntimes;
+    }
+
+    public IReadOnlyCollection<TestConfiguration> CreateTestConfigurations (
+        string testAssemblyFullPath,
+        string semicolonSeparatedTestConfigurations,
+        string testSetupBuildFile = null)
+    {
+      EnsureTestConfigurationsExist (testAssemblyFullPath, semicolonSeparatedTestConfigurations);
+
+      _supportedExecutionRuntimeDictionary = GetSupportedExecutionRuntimeDictionary (_supportedExecutionRuntimes);
+
+      var output = new List<TestConfiguration>();
+
+      var plusSeparatedConfigurationItemsCollection = GetPlusSeparatedConfigurations (semicolonSeparatedTestConfigurations);
+      foreach (var plusSeparatedConfigurationItems in plusSeparatedConfigurationItemsCollection)
+      {
+        var configurationItems = GetConfigurationItems (plusSeparatedConfigurationItems);
+
+        EnsureNoDuplicateItemsExist (configurationItems);
+
+        var testConfiguration = CreateTestConfiguration (
+            testAssemblyFullPath,
+            configurationItems,
+            plusSeparatedConfigurationItems,
+            testSetupBuildFile);
+
+        output.Add (testConfiguration);
+      }
+
+      return output;
+    }
+
+    private void EnsureTestConfigurationsExist (string testAssemblyFullPath, string semicolonSeparatedTestConfigurations)
+    {
+      if (string.IsNullOrEmpty (semicolonSeparatedTestConfigurations))
+        throw new ArgumentException ($"Could not find TestConfigurations for {testAssemblyFullPath}.");
+    }
+
+    private Dictionary<string, string> GetSupportedExecutionRuntimeDictionary (IReadOnlyCollection<string> supportedExecutionRuntimes)
+    {
+      return supportedExecutionRuntimes
+          .Select (x => x.Split (new[] { "::" }, StringSplitOptions.RemoveEmptyEntries))
+          .ToDictionary (split => split[0], split => split[1], StringComparer.OrdinalIgnoreCase);
+    }
+
+    private IReadOnlyCollection<string> GetPlusSeparatedConfigurations (string testingConfiguration)
+    {
+      return Regex.Replace (testingConfiguration, @"\s+", "").Split (new[] { ";" }, StringSplitOptions.RemoveEmptyEntries);
+    }
+
+    private IReadOnlyCollection<string> GetConfigurationItems (string configuration)
+    {
+      return Regex.Replace (configuration, @"\s+", "").Split ('+');
+    }
+
+    private void EnsureNoDuplicateItemsExist (IReadOnlyCollection<string> configurationItems)
+    {
+      var duplicateItems = GetDuplicateItems (configurationItems);
+
+      if (duplicateItems.Any())
+      {
+        throw new InvalidOperationException ($"The following configuration values were found multiple times: '{string.Join (",", duplicateItems)}'");
```

Comment:
I thought `Ensure` would be better than `Check` because if the check fails if throws an exception and only `Check` implies more of a "let's return a bool" kind of deal. What's the naming convention here if we not only check but actually stop the program from continuation if the check fails? Wouldn't it even be `Verify` or something?

Reply:
https://github.com/re-motion/Trunk-SVN2Git/blob/master/SharedSource/Core/Utilities/ArgumentUtility.cs#L57